hnn_core/extracellular.py

```diff
@@ -0,0 +1,551 @@
+            Labels associated with the contacts to plot. Passed as-is to
+            :func:`~matplotlib.axes.Axes.set_yticklabels`.
+        show : bool
+            If True, show the figure
+
+        Returns
+        -------
+        fig : instance of plt.fig
+            The matplotlib figure handle into which time series were plotted.
+        """
+        from .viz import plot_extracellular
+
+        if trial_no is None:
+            plot_data = self.get_data()
+        elif isinstance(trial_no, (list, tuple, int, slice)):
+            plot_data = self.get_data()[trial_no, ]
+        else:
+            raise ValueError(f'unkown trial number type, got {trial_no}')
+
+        if isinstance(contact_no, (list, tuple, int, slice)):
+            plot_data = plot_data[:, contact_no, ]
+        elif contact_no is not None:
+            raise ValueError(f'unkown contact number type, got {contact_no}')
+
+        for trial_data in plot_data:
+            fig = plot_extracellular(
+                self.times, trial_data, self.sfreq, window_len=window_len,
+                tmin=tmin, tmax=tmax, ax=ax, decim=decim, color=color,
+                voltage_offset=voltage_offset,
+                voltage_scalebar=voltage_scalebar,
+                contact_labels=contact_labels,
+                show=show)
+        return fig
+
+
+class _ExtracellularArray:
+    """Class for electrode arrays containing NEURON objects
+
+    The handler is set up to maintain a vector of membrane currents at at every
+    inner segment of every section of every cell on each CVODE integration
+    step. In addition, it records a time vector of sample times. This class
+    must be instantiated and attached to the network during the building
+    process. It is used in conjunction with the calculation of extracellular
+    potentials.
+
+    Parameters
+    ----------
+    array : instance of ExtracellularArray
+        Initialised, e.g., by the `Network.add_electrode_array()`-method
+    cvode : instance of h.CVode
+        Multi order variable time step integration method.
+    """
+    def __init__(self, array, cvode=None):
+        self.positions = array.positions
+        self.n_contacts = len(self.positions)
+        self.sigma = array.sigma
+        self.min_distance = array.min_distance
+        self.method = array.method
+
+        # Create the Neuron objects needed to record extracellular voltages
+
+        # ordered list of h.Sections on this rank (if running in parallel)
+        secs_on_rank = _get_sections_on_this_rank()
+        # np.array of number of segments for each section, ordered as above
+        segment_counts = np.array(_get_segment_counts(secs_on_rank))
+
+        # pointers assigned to _ref_i_membrane_ at each EACH internal segment
+        self.imem_ptrvec = h.PtrVector(segment_counts.sum())
+        # placeholder into which pointer values are read on each sim step
+        self.imem_vec_len = int(self.imem_ptrvec.size())
```

Comment:
is `imem_vec_len` = `n_times`. The name is a bit ambiguous as it stands

Reply:
No, it's the length of the vector that contains a single time instant of the transmembrane current at each of the segments (on this rank). The values are gathered from `self.imem_ptrvec` at each time step and multiplied by the transfer resistances to each electrode.